proposals/type-applications-in-patterns.rst

```diff
@@ -0,0 +1,170 @@
+
+If ``ScopedTypeVariables`` is enabled, then type applications and type signatures in pattern may mention type variables.
+
+A type variable mentioned in a pattern (in a type signature or a type application) that is not already in scope is brought into scope. The scope spans the pattern and the corresponding right-hand side.
+
+Typechecking a type application in a pattern follows the same rules as typechecking a pattern signature. This proposal does not propose any changes to the mechanism. In particular
+
+* type variables in type applications in patterns can only bind type variables, just as it is the case for type variables in pattern signatures (unless `Proposal #128 <https://github.com/ghc-proposals/ghc-proposals/pull/128>`_ gets accepted first).
+* patterns are type-checked from left to right. This means when type-checking the type application in a pattern ``(MkT @ty)``, type equatlities from matches further to the left, as well as the type equalities from the context of ``MkT`` itself are in scope, but not type equalities from constructors further to the right (see below for an example).
+
+A more formal description of these rules can be found in `“Type variable in pattern” by Richard Eisenberg, Joachim Breitner  and Simon Peyton Jones <https://arxiv.org/abs/1806.03476>`_.
+
+An underscore in a type signature or type application in a pattern is not treated as a hole.
+
+Examples
+--------
+
+Here is an example (taken from _#15050 <https://ghc.haskell.org/trac/ghc/ticket/15050#comment:10>_)::
+
+    type family F a where F Bool = Int
+    data T a where MkT :: forall b a. b ~ F a => b -> T a
+    foo :: T Bool -> ()
+    foo (MkT @Int _) = ()
+
+This should type-check, because the following code does::
+
+    foo :: T Bool -> ()
+    foo (MkT (_ ::Int _)) = ()
+
+
+A more complex example is this (also inspired by _#15050 <https://ghc.haskell.org/trac/ghc/ticket/15050>_)::
+
+    data T a where
+      MkT1 :: forall a.              T a
+      MkT2 :: forall a.              T (a,a)
+      MkT3 :: forall a b. b ~ Int => T a
+      MkT4 :: forall a b.            T a
+      MkT5 :: forall a b c. b ~ c => T a
+      
+    foo :: T (Int, Int) -> ()
+    foo (MkT1 @(Int,Int))  = ()
+    foo (MkT2 @x)          = (() :: x ~ Int => ())
+    foo (MkT3 @_ @Int)     = ()
+    foo (MkT4 @_ @x)       = (() :: x ~ x => ()) -- (these constraints here just to
+    foo (MkT5 @_ @x @x)    = (() :: x ~ x => ()) --  demonstrate that x is in scope)
+
+All of these equations type-check (just like they would if added value arguments of type ``a``, ``b``,... to the constructors and turned the type applications into type signatures).
+
+This example demonstrated why we need to typecheck nested patterns left-to-right::
+
+ data T a where
+   T1 :: T Int
+   T2 :: T a
+
+ f :: Int -> Char -> Bool
+
+ g :: (a, Char, T a) -> blah
+ g (x :: Int, (f x -> True), T1) = ..
+
+``g`` must not be accepted: Until we match on ``T1`` we have no idea if ``a ~ Int``.
```

Comment:
This is absolutely nothing to do with visible type application in patterns!  What role does this example play in the proposal??

Reply:
Removed it.